libnavigation-base/src/main/java/com/mapbox/navigation/base/utils/route/NavigationRouteEx.kt

```diff
@@ -0,0 +1,170 @@
+                }
+                if (closure.geometryIndexStart() != 0 &&
+                    closure.geometryIndexEnd() != legLastGeometryIndex &&
+                    silentWaypointsInClosureRange.isEmpty()
+                ) {
+                    logD(
+                        "Route with id [${this@hasUnexpectedClosures.id}] has closure at leg " +
+                            "index $routeLegIndex",
+                    )
+                    return@withContext true
+                }
+                if (closure.geometryIndexStart() == 0 &&
+                    snappingResultList.getOrNull(currentLegFirstWaypointIndex) != true
+                ) {
+                    logD(
+                        "Route with id [${this@hasUnexpectedClosures.id}] has closure at the " +
+                            "start of the leg, leg index $routeLegIndex",
+                        LOG_CATEGORY
+                    )
+                    return@withContext true
+                }
+                if (closure.geometryIndexEnd() == legLastGeometryIndex &&
+                    snappingResultList.getOrNull(
+                            currentLegFirstWaypointIndex + silentWaypoints.size + 1
+                        ) != true
+                ) {
+                    logD(
+                        "Route with id [${this@hasUnexpectedClosures.id}] has closure at " +
+                            "the end of the leg, leg index $routeLegIndex",
+                        LOG_CATEGORY
+                    )
+                    return@withContext true
+                }
+            }
+            currentLegFirstWaypointIndex += silentWaypoints.size
+            currentLegFirstWaypointIndex++
+        }
+
+        return@withContext false
+    }
+
+private fun DirectionsRoute.getSnappingResultList(): List<Boolean> {
+    val snappingIncludeClosuresList = routeOptions()?.snappingIncludeClosuresList()
+    val snappingIncludeStaticClosuresList = routeOptions()?.snappingIncludeStaticClosuresList()
+
+    val snappingResultList = mutableListOf<Boolean>()
+    for (
+        index in 0 until maxOf(
+            snappingIncludeClosuresList?.size ?: 0,
+            snappingIncludeStaticClosuresList?.size ?: 0
+        )
+    ) {
+        snappingResultList.add(
+            snappingIncludeClosuresList?.getOrNull(index) ?: false ||
+                snappingIncludeStaticClosuresList?.getOrNull(index) ?: false
+        )
+    }
+    return snappingResultList
+}
+
+/**
+ * return the list of indexes of List<LegSilentWaypoints> which are at closure area
+ */
+private fun List<LegSilentWaypoints>.inGeometryRange(
+    geometryIndexStart: Int,
+    geometryIndexEnd: Int
+): List<Int> {
+    return mapIndexedNotNull { index, legSilentWaypoints ->
+        if (legSilentWaypoints.geometryIndex in geometryIndexStart..geometryIndexEnd) {
+            index
+        } else {
+            null
+        }
+    }
+}
+
+private fun DirectionsRoute.stepsGeometryToPoints(
```

Comment:
Actually, I use it, see the line `stepGeometryToPoints(legStep).let {`. I needed to clean up duplicated `Point`s. It would be nice to move this to extensions of `DecodeUtils`. Let's address it separately when we get that it works good